dataclients/kubernetes/kube.go

```diff
@@ -802,217 +817,258 @@ func (c *Client) ingressToRoutes(state *clusterState, defaultFilters map[resourc
 	hostRoutes := make(map[string][]*eskip.Route)
 	redirect := createRedirectInfo(c.provideHTTPSRedirect, c.httpsRedirectCode)
 	for _, i := range state.ingresses {
-		if i.Metadata == nil || i.Metadata.Namespace == "" || i.Metadata.Name == "" ||
-			i.Spec == nil {
-			log.Error("invalid ingress item: missing metadata")
-			continue
+		r, err := c.ingressRoute(i, redirect, state, hostRoutes, defaultFilters)
+		if err != nil {
+			return nil, err
 		}
-
-		logger := log.WithFields(log.Fields{
-			"ingress": fmt.Sprintf("%s/%s", i.Metadata.Namespace, i.Metadata.Name),
-		})
-
-		redirect.initCurrent(i.Metadata)
-
-		if r, ok, err := c.convertDefaultBackend(state, i); ok {
+		if r != nil {
 			routes = append(routes, r)
-		} else if err != nil {
-			logger.Errorf("error while converting default backend: %v", err)
 		}
+	}
 
-		// parse filter and ratelimit annotation
-		var annotationFilter string
-		if ratelimitAnnotationValue, ok := i.Metadata.Annotations[ratelimitAnnotationKey]; ok {
-			annotationFilter = ratelimitAnnotationValue
+	for host, rs := range hostRoutes {
+		if len(rs) == 0 {
+			continue
 		}
-		if val, ok := i.Metadata.Annotations[skipperfilterAnnotationKey]; ok {
-			if annotationFilter != "" {
-				annotationFilter += " -> "
-			}
-			annotationFilter += val
+
+		routes = append(routes, rs...)
+
+		// if routes were configured, but there is no catchall route
+		// defined for the host name, create a route which returns 404
+		if !catchAllRoutes(rs) {
+			routes = append(routes, c.addCatchAllRoutes(host, rs[0], redirect)...)
 		}
+	}
 
-		// parse predicate annotation
-		var annotationPredicate string
-		if val, ok := i.Metadata.Annotations[skipperpredicateAnnotationKey]; ok {
-			annotationPredicate = val
+	return routes, nil
+}
+
+func (c *Client) ingressRoute(i *ingressItem, redirect *redirectInfo, state *clusterState,
```

Comment:
while extracting this block to a method looks safe, it still could be cleaner. What makes a bit hard to comprehend is that it returns one route, but potentially it creates more routes which are added to the hostRoutes argument as a side effect.